File: python/legacy_tax_calculator.py
"""Chapter 9: a line-for-line port of legacyTaxCalculator.ts."""
import math


def _js_round(x: float) -> int:
    # JavaScript's Math.round (half up), not Python's round().
    return math.floor(x + 0.5)


def calc_tax(o: dict, st: str, cty: str | None = None) -> int:
    r = 0.0
    if st == "CA":
        r = 0.0725
        if cty == "LA":
            r += 0.01
        if o.get("type") == "food" and o.get("roasted") is not True:
            r = 0
    elif st == "CO":
        r = 0.029
        if cty == "Boulder" and o.get("type") == "food":
            r = 0
    elif st == "NY":
        r = 0.08875 if o["subtotal"] > 11000 else 0.04
    # TODO: fix this for subscriptions, see ticket PARC-448
    if o.get("isSubscription") and st != "NY":
        r = r * 0.5
    return _js_round(o["subtotal"] * r)

File: src/tax/legacyTaxCalculator.ts
// Chapter 9: the file nobody wants to touch. Reproduced as-is,
// warts included; the tests characterize it, they don't bless it.
export function calcTax(o: any, st: string, cty?: string) {
  let r = 0;
  if (st === "CA") {
    r = 0.0725;
    if (cty === "LA") r += 0.01;
    if (o.type === "food" && o.roasted !== true) r = 0;
  } else if (st === "CO") {
    r = 0.029;
    if (cty === "Boulder" && o.type === "food") r = 0;
  } else if (st === "NY") {
    r = o.subtotal > 11000 ? 0.08875 : 0.04;
  }
  // TODO: fix this for subscriptions, see ticket PARC-448
  if (o.isSubscription && st !== "NY") r = r * 0.5;
  return Math.round(o.subtotal * r);
}
